### PR TITLE
Implement kNN for tf.idf similarity and use it for prediction

### PR DESCRIPTION
**Pull request by blanche**
also
- separate vectorization from similarity/knn computation
- use stopwords & lowercase in vectorizer
- use the same random seed for every model (still random, but better for comparison)
- output model name when printing error

### CODE REVIEW

**Comment by utkuozbulak** [comment]
Can you please not merge this just for a little more while, I we, I will check it as soon as possible.

**Comment by blanche** [comment]
sure!
no rush for this one.. my next task is cleaning up the companies anyway

btw, the knn thing its pretty fast, even on 100k data set (i.e. < 5 min?), but i get memory error on my machine when running it on the full 120k, maybe have to play with the parallelization..

**Comment by blanche** [comment]
done
error = 7.4k

**Comment by utkuozbulak** [comment]
on full dataset ?

File: src/models/BaseModel.py [hunk 1]
```diff
@@ -38,6 +39,11 @@ def __init__(self, train_size=0.75, test_size=None):
             print('Pre-processed data exists, reading from the file')
             self.cleaned_encoded_data = self.load_cleaned_data()
             print('Data read complete')
+            if force_load_all:
```
Comment (arahayrabedian) [inline src/models/BaseModel.py:42]: could you summarize what's happening here? i'm a little lost as to why only `self.description_feature` is re-read.

File: src/models/StandaloneSimilarity.py [hunk 2]
```diff
@@ -1,27 +1,169 @@
...
-"""
+from cleaning_functions import pandas_vector_to_list
+from job_description_feature_extraction import (tfidf_vectorize,
+                                                cosine_knn, calc_cosine_sim,
+                                                extract_top_k_documents)
 from models.BaseModel import BaseModel
 
+DATA_QUERIES_VECTOR_NPZ = '../preprocessing_data/queries_vector.npz'
+DATA_CORPUS_VECTOR_NPZ = '../preprocessing_data/corpus_vector.npz'
+DATA_SIMILIARITY_MATRIX = '../preprocessing_data/similiarity_matrix'
```
Comment (arahayrabedian) [inline src/models/StandaloneSimilarity.py:19]: `.npz`? unused?

File: src/models/StandaloneSimilarity.py [hunk 3]
```diff
@@ -1,27 +1,169 @@
...
+DATA_Y_TRAIN = '../preprocessing_data/y_train'
+DATA_Y_TEST = '../preprocessing_data/y_test'
+
+FORCE_LOAD = True
+
 
 class StandaloneSimilarity(BaseModel):
```
Comment (arahayrabedian) [inline src/models/StandaloneSimilarity.py:26]: i'mma just trust you on this chunk of code.

File: src/models/StandaloneSimilarity.py [hunk 4]
```diff
@@ -1,27 +1,169 @@
...
+        # # multiple versions of calculating the prediction, some faster, some use more mem
+
+        # prediction = self.multiprocessor_batch_calc(batch_queue)
+        prediction = self.batch_calculation(batch_queue)
+        # prediction = self.individual_calculation()
+        # prediction = self.cosine_knn_calc()
+        # prediction = self.custom_knn_calculation(prediction)
+
+        train_avg_salary = sum(self.y_train) / len(self.y_train)
+        cleaned_predictions = [x if str(x) != 'nan' else train_avg_salary for x in prediction]
```
Comment (arahayrabedian) [inline src/models/StandaloneSimilarity.py:55]: there's got to be a better way to check for NaN

Comment (arahayrabedian) [reply src/models/StandaloneSimilarity.py:55]: also - average if nan? safe assumption? when does this happen?

File: src/models/BaseModel.py (hunk 1 above)
Comment (blanche) [reply src/models/BaseModel.py:42]: the problem is, that the `cleaned_encoded_data` does not contain the raw description data we need for the similarity model here. 

all other models are fine with the cleaned_encoded_data, so using the force_load_all you can force it to load (at least) the raw description data

**Review by arahayrabedian — COMMENTED** [review]
not managing to internalize everything going on here, but looks generally not so bad

File: src/models/StandaloneSimilarity.py (hunk 2 above)
Comment (blanche) [reply src/models/StandaloneSimilarity.py:19]: npz is a numpy format
see https://docs.scipy.org/doc/numpy/reference/generated/numpy.savez.html

its very much used in models.StandaloneSimilarity.StandaloneSimilarity#store_precomputed_data

File: src/models/StandaloneSimilarity.py (hunk 4 above)
Comment (blanche) [reply src/models/StandaloneSimilarity.py:55]: tbh i don't understand why other models don't have that problem, but it happens if you predict on the basis of other Nans (similar to having a 0 in a chain of multiplications)

better solution might be to clean input data better - should i give that a try?

if we cant make a proper prediction, we just fall back to the average salary

that being said i have only seen a single case of this in the entire data set - but the error calculation will fail if the predictions contain a nan value

**Review by alexdy2007 — APPROVED** [review]
